src/main/java/com/puppycrawl/tools/checkstyle/utils/CheckUtil.java

```diff
@@ -484,4 +489,95 @@ else if (tokenType == TokenTypes.LITERAL_PRIVATE) {
         return illegalClassNames;
     }
 
+    /**
+     * Strip initial newline and preceding whitespace on each line from text block content.
+     * In order to be consistent with how javac handles this task, we have modeled this
+     * implementation after the code from:
+     * github.com/openjdk/jdk14u/blob/master/src/java.base/share/classes/java/lang/String.java
+     *
+     * @param textBlockContent the actual content of the text block.
+     * @return string consistent with javac representation.
+     */
+    public static String stripIndentAndInitialNewLineFromTextBlock(String textBlockContent) {
+        final String contentWithInitialNewLineRemoved =
+            ALL_NEW_LINES.matcher(textBlockContent).replaceFirst("");
+        final List<String> lines =
+            Arrays.asList(ALL_NEW_LINES.split(contentWithInitialNewLineRemoved));
+        final int indent = getSmallestIndent(lines);
+        final String suffix = "";
+
+        return lines.stream()
+                .map(line -> stripIndentAndTrailingWhitespaceFromLine(line, indent))
+                .collect(Collectors.joining(System.lineSeparator(), "", suffix));
+    }
+
+    /**
+     * Helper method for stripIndentAndInitialNewLineFromTextBlock, strips correct indent
+     * from string, and trailing whitespace, or returns empty string if no text.
+     *
+     * @param line the string to strip indent and trailing whitespace from
+     * @param indent the amount of indent to remove
+     * @return modified string with removed indent and trailing whitespace, or empty string.
+     */
+    private static String stripIndentAndTrailingWhitespaceFromLine(String line, int indent) {
+        final int lastNonWhitespace = lastIndexOfNonWhitespace(line);
+        String returnString = "";
+
+        if (lastNonWhitespace > 0) {
+            final int firstNonWhitespace = indexOfNonWhitespace(line);
+            final int incidentalWhitespace = Math.min(indent, firstNonWhitespace);
+            returnString = line.substring(incidentalWhitespace, lastNonWhitespace);
+        }
```

Comment:
There should be `else return ""';` as the compiler threats whitespace-only lines as empty:
```
static String str = """
          
   foo""";
   System.out.println("'" + str + "'");
```
(there is 10 spaces in the line 2) becomes
```
'
foo'
```

Reply:
https://github.com/checkstyle/checkstyle/pull/8705/files/ac210211923bf776a214945259bd63e501b363cd#r473605759